// darn-simple-audio/SingleAudioPlayer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Controls = _interopRequireDefault(require("./components/Controls"));

var _CurrentTrack = _interopRequireDefault(require("./components/CurrentTrack"));

var _propTypes = _interopRequireDefault(require("prop-types"));

require("./assets/style.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var SingleAudioPlayer =
/*#__PURE__*/
function (_Component) {
  _inherits(SingleAudioPlayer, _Component);

  function SingleAudioPlayer(props) {
    var _this;

    _classCallCheck(this, SingleAudioPlayer);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SingleAudioPlayer).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "setAudioElement", function (e) {
      _this.setState({
        audioElement: e
      }, function () {
        if (_this.props.autoPlay) {
          _this.playAudio();
        }
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "setTotalDuration", function (e) {
      if (_this.props.onTotalDurationChange && typeof _this.props.onTotalDurationChange === 'function') {
        _this.props.onTotalDurationChange(e.target.duration);
      }

      _this.setState({
        totalDuration: e.target.duration
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "setEnded", function (e) {
      if (e.target.ended) {
        _this.setState({
          endReached: e.target.ended,
          playing: false
        });
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "setPlaying", function () {
      _this.setState({
        playing: true,
        endReached: false
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "setCurrentPlayTime", function (e) {
      if (_this.props.onCurrentTimeChange && typeof _this.props.onCurrentTimeChange === 'function') {
        _this.props.onCurrentTimeChange(e.target.currentTime);
      }

      _this.setState({
        currentPlayTime: e.target.currentTime
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "playAudio", function () {
      _this._playAudioFrom(0);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_playAudioFrom", function (t) {
      var currentState = _objectSpread({}, _this.state);

      currentState.audioElement.currentTime = t;
      currentState.currentPlayTime = t;

      _this.setState(currentState, function () {
        if (_this.state.audioElement.load) {
          if (t === 0) {
            _this.state.audioElement.load();
          }

          _this.state.audioElement.play();
        }
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleClick", function () {
      console.error("Navigating between tracks is not yet supported.");
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "resumeAudio", function () {
      _this._playAudioFrom(_this.state.currentPlayTime);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "pauseAudio", function () {
      _this.setState({
        playing: false
      });

      _this.state.audioElement.pause();
    });

    _this.state = {
      playing: false,
      audioElement: {},
      currentPlayTime: 0,
      totalDuration: 0,
      endReached: false
    };
    return _this;
  }

  _createClass(SingleAudioPlayer, [{
    key: "render",
    value: function render() {
      return _react.default.createElement("div", {
        className: "darn-simple-audio-container"
      }, _react.default.createElement("div", {
        className: "album-art",
        style: {
          backgroundImage: "url(" + this.props.albumArt + ")",
          width: this.props.width,
          height: this.props.height
        }
      }, this.props.showControls ? _react.default.createElement(_Controls.default, _extends({
        onClick: this.handleClick,
        resumeAudio: this.resumeAudio,
        pause: this.pauseAudio,
        play: this.playAudio
      }, this.state, this.props)) : '', this.props.showTrackTitle ? _react.default.createElement("span", null, this.props.trackTile) : '', _react.default.createElement(_CurrentTrack.default, _extends({
        setCurrentPlayTime: this.setCurrentPlayTime,
        setTotalDuration: this.setTotalDuration,
        play: this.playAudio,
        ended: this.setEnded,
        setAudioElement: this.setAudioElement,
        setEnded: this.setEnded,
        setPlaying: this.setPlaying
      }, this.props))));
    }
  }]);

  return SingleAudioPlayer;
}(_react.Component);

SingleAudioPlayer.defaultProps = {
  showControls: true,
  showPrev: false,
  showNext: false,
  showTrackTitle: false,
  autoPlay: true,
  showReplay: true,
  showPause: true,
  width: 280,
  height: 280,
  source: "",
  showAlbumTitle: true,
  albumArt: ""
};
SingleAudioPlayer.propTypes = {
  showControls: _propTypes.default.bool,
  showPrev: _propTypes.default.bool,
  showNext: _propTypes.default.bool,
  autoPlay: _propTypes.default.bool,
  showPause: _propTypes.default.bool,
  showReplay: _propTypes.default.bool,
  showAlbumTitle: _propTypes.default.bool,
  trackTile: _propTypes.default.string,
  source: _propTypes.default.string,
  albumArt: _propTypes.default.string,
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  height: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  onCurrentTimeChange: _propTypes.default.func,
  onTotalDurationChange: _propTypes.default.func,
  prevIcon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.element]),
  nextIcon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.element]),
  playIcon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.element]),
  pauseIcon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.element]),
  replayIcon: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.element])
};
var _default = SingleAudioPlayer;
exports.default = _default;
